File: mediapipe/tasks/javatests/com/google/mediapipe/tasks/vision/facestylizer/FaceStylizerTest.java
// Copyright 2023 The MediaPipe Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.mediapipe.tasks.vision.facestylizer;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.Assert.assertThrows;

import android.content.res.AssetManager;
import android.graphics.BitmapFactory;
import android.graphics.RectF;
import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import com.google.mediapipe.framework.image.BitmapImageBuilder;
import com.google.mediapipe.framework.image.MPImage;
import com.google.mediapipe.tasks.core.BaseOptions;
import com.google.mediapipe.tasks.vision.core.ImageProcessingOptions;
import com.google.mediapipe.tasks.vision.facestylizer.FaceStylizer.FaceStylizerOptions;
import java.io.InputStream;
import org.junit.After;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Suite;
import org.junit.runners.Suite.SuiteClasses;

/** Test for {@link FaceStylizer}. */
@RunWith(Suite.class)
@SuiteClasses({FaceStylizerTest.General.class, FaceStylizerTest.RunningModeTest.class})
public class FaceStylizerTest {
  private static final String modelFile = "face_stylizer.task";
  private static final String largeFaceTestImage = "portrait.jpg";
  private static final String smallFaceTestImage = "portrait_small.jpg";
  private static final int modelImageSize = 256;

  @RunWith(AndroidJUnit4.class)
  public static final class General extends FaceStylizerTest {
    FaceStylizer faceStylizer;

    @After
    public void afterEach() throws Exception {
      if (faceStylizer != null) {
        faceStylizer.close();
      }
    }

    @Test
    public void create_succeeds() throws Exception {
      FaceStylizerOptions options =
          FaceStylizerOptions.builder()
              .setBaseOptions(BaseOptions.builder().setModelAssetPath(modelFile).build())
              .build();
      faceStylizer =
          FaceStylizer.createFromOptions(ApplicationProvider.getApplicationContext(), options);
      assertThat(faceStylizer).isNotNull();
    }

    @Test
    public void create_failsWithMissingModel() throws Exception {
      IllegalArgumentException exception =
          assertThrows(
              IllegalArgumentException.class,
              () ->
                  FaceStylizerOptions.builder()
                      .setBaseOptions(BaseOptions.builder().build())
                      .build());
      assertThat(exception)
          .hasMessageThat()
          .contains(
              "specify only one of the model asset path, the model asset file descriptor, and the"
                  + " model asset buffer");
    }
  }

  @RunWith(AndroidJUnit4.class)
  public static final class RunningModeTest extends FaceStylizerTest {
    FaceStylizer faceStylizer;

    @After
    public void afterEach() throws Exception {
      if (faceStylizer != null) {
        faceStylizer.close();
      }
    }

    @Test
    public void stylizer_succeedsWithImageMode() throws Exception {
      FaceStylizerOptions options =
          FaceStylizerOptions.builder()
              .setBaseOptions(BaseOptions.builder().setModelAssetPath(modelFile).build())
              .build();

      faceStylizer =
          FaceStylizer.createFromOptions(ApplicationProvider.getApplicationContext(), options);

      MPImage inputImage = getImageFromAsset(largeFaceTestImage);
      FaceStylizerResult actualResult = faceStylizer.stylize(inputImage);
      MPImage stylizedImage = actualResult.stylizedImage().get();
      assertThat(stylizedImage).isNotNull();
      assertThat(stylizedImage.getWidth()).isEqualTo(modelImageSize);
      assertThat(stylizedImage.getHeight()).isEqualTo(modelImageSize);
    }

    @Test
    public void stylizer_succeedsWithSmallImage() throws Exception {
      FaceStylizerOptions options =
          FaceStylizerOptions.builder()
              .setBaseOptions(BaseOptions.builder().setModelAssetPath(modelFile).build())
              .build();

      faceStylizer =
          FaceStylizer.createFromOptions(ApplicationProvider.getApplicationContext(), options);

      MPImage inputImage = getImageFromAsset(smallFaceTestImage);
      FaceStylizerResult actualResult = faceStylizer.stylize(inputImage);
      MPImage stylizedImage = actualResult.stylizedImage().get();
      assertThat(stylizedImage).isNotNull();
      assertThat(stylizedImage.getWidth()).isEqualTo(modelImageSize);
      assertThat(stylizedImage.getHeight()).isEqualTo(modelImageSize);
    }

    @Test
    public void stylizer_succeedsWithRegionOfInterest() throws Exception {
      FaceStylizerOptions options =
          FaceStylizerOptions.builder()
              .setBaseOptions(BaseOptions.builder().setModelAssetPath(modelFile).build())
              .build();
      faceStylizer =
          FaceStylizer.createFromOptions(ApplicationProvider.getApplicationContext(), options);

      MPImage inputImage = getImageFromAsset(largeFaceTestImage);

      // Region-of-interest around the face.
      RectF roi =
          new RectF(/* left= */ 0.32f, /* top= */ 0.02f, /* right= */ 0.67f, /* bottom= */ 0.32f);
      ImageProcessingOptions imageProcessingOptions =
          ImageProcessingOptions.builder().setRegionOfInterest(roi).build();

      FaceStylizerResult actualResult = faceStylizer.stylize(inputImage, imageProcessingOptions);
      MPImage stylizedImage = actualResult.stylizedImage().get();
      assertThat(stylizedImage).isNotNull();
      assertThat(stylizedImage.getWidth()).isEqualTo(modelImageSize);
      assertThat(stylizedImage.getHeight()).isEqualTo(modelImageSize);
    }

    @Test
    public void stylizer_succeedsWithNoFaceDetected() throws Exception {
      FaceStylizerOptions options =
          FaceStylizerOptions.builder()
              .setBaseOptions(BaseOptions.builder().setModelAssetPath(modelFile).build())
              .build();
      faceStylizer =
          FaceStylizer.createFromOptions(ApplicationProvider.getApplicationContext(), options);

      MPImage inputImage = getImageFromAsset(largeFaceTestImage);

      // Region-of-interest that doesn't contain a human face.
      RectF roi =
          new RectF(/* left= */ 0.1f, /* top= */ 0.1f, /* right= */ 0.2f, /* bottom= */ 0.2f);
      ImageProcessingOptions imageProcessingOptions =
          ImageProcessingOptions.builder().setRegionOfInterest(roi).build();
      FaceStylizerResult actualResult = faceStylizer.stylize(inputImage, imageProcessingOptions);
      assertThat(actualResult.stylizedImage()).isNotNull();
      assertThat(actualResult.stylizedImage().isPresent()).isFalse();
    }

    @Test
    public void stylizer_successWithImageModeWithResultListener() throws Exception {
      MPImage inputImage = getImageFromAsset(largeFaceTestImage);
      FaceStylizerOptions options =
          FaceStylizerOptions.builder()
              .setBaseOptions(BaseOptions.builder().setModelAssetPath(modelFile).build())
              .setResultListener(
                  (result, originalImage) -> {
                    MPImage stylizedImage = result.stylizedImage().get();
                    assertThat(stylizedImage).isNotNull();
                    assertThat(stylizedImage.getWidth()).isEqualTo(modelImageSize);
                    assertThat(stylizedImage.getHeight()).isEqualTo(modelImageSize);
                  })
              .build();
      faceStylizer =
          FaceStylizer.createFromOptions(ApplicationProvider.getApplicationContext(), options);
      faceStylizer.stylizeWithResultListener(getImageFromAsset(largeFaceTestImage));
    }
  }

  private static MPImage getImageFromAsset(String filePath) throws Exception {
    AssetManager assetManager = ApplicationProvider.getApplicationContext().getAssets();
    InputStream istr = assetManager.open(filePath);
    return new BitmapImageBuilder(BitmapFactory.decodeStream(istr)).build();
  }
}

File: mediapipe/tasks/python/test/metadata/metadata_writers/face_stylizer_test.py
# Copyright 2023 The MediaPipe Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Tests for metadata_writer.face_stylizer."""

import os
import tempfile
import zipfile

from absl.testing import absltest
from absl.testing import parameterized

from mediapipe.tasks.python.metadata.metadata_writers import face_stylizer
from mediapipe.tasks.python.test import test_utils

_TEST_DATA_DIR = "mediapipe/tasks/testdata/metadata"
_NORM_MEAN = 0
_NORM_STD = 255
_TFLITE = test_utils.get_test_data_path(
    os.path.join(_TEST_DATA_DIR, "dummy_face_stylizer.tflite")
)
_EXPECTED_JSON = test_utils.get_test_data_path(
    os.path.join(_TEST_DATA_DIR, "face_stylizer.json")
)


class FaceStylizerTest(parameterized.TestCase):

  def test_write_metadata_and_create_model_asset_bundle_successful(self):
    # Use dummy model buffer for unit test only.
    with open(_TFLITE, "rb") as f:
      face_stylizer_model_buffer = f.read()
    face_detector_model_buffer = b"\x33\x44"
    face_landmarks_detector_model_buffer = b"\x55\x66"
    writer = face_stylizer.MetadataWriter.create(
        face_stylizer_model_buffer,
        face_detector_model_buffer,
        face_landmarks_detector_model_buffer,
        input_norm_mean=[_NORM_MEAN],
        input_norm_std=[_NORM_STD],
    )
    model_bundle_content, metadata_json = writer.populate()
    with open(_EXPECTED_JSON, "r") as f:
      expected_json = f.read()
    self.assertEqual(metadata_json, expected_json)

    temp_folder = tempfile.TemporaryDirectory()

    # Checks the model bundle can be extracted successfully.
    model_bundle_filepath = os.path.join(temp_folder.name, "face_stylizer.task")

    with open(model_bundle_filepath, "wb") as f:
      f.write(model_bundle_content)

    with zipfile.ZipFile(model_bundle_filepath) as zf:
      self.assertEqual(
          set(zf.namelist()),
          set([
              "face_detector.tflite",
              "face_landmarks_detector.tflite",
              "face_stylizer.tflite",
          ]),
      )
      zf.extractall(temp_folder.name)
    temp_folder.cleanup()


if __name__ == "__main__":
  absltest.main()
